provisioner/node_pools.go

```diff
@@ -53,46 +53,111 @@ type remoteData struct {
 	} `yaml:"files" json:"files"`
 }
 
-// AWSNodePoolProvisioner is a node provisioner able to provision node pools
-// in AWS via cloudformation.
-// TODO: move AWS specific implementation to a separate file/package.
-type AWSNodePoolProvisioner struct {
-	awsAdapter      *awsAdapter
-	instanceTypes   *awsUtils.InstanceTypes
-	nodePoolManager updatestrategy.NodePoolManager
-	bucketName      string
-	config          channel.Config
-	cluster         *api.Cluster
-	azInfo          *AZInfo
-	templateContext *templateContext
-	logger          *log.Entry
+type NodePoolTemplateRenderer struct {
+	awsAdapter     *awsAdapter
+	config         channel.Config
+	cluster        *api.Cluster
+	bucketName     string
+	logger         *log.Entry
+	encodeUserData bool
 }
 
-func (p *AWSNodePoolProvisioner) generateNodePoolStackTemplate(nodePool *api.NodePool, values map[string]interface{}) (string, error) {
-	renderedUserData, err := p.prepareCloudInit(nodePool, values)
+// prepareCloudInit prepares the user data by rendering the golang template.
+// It also uploads the dynamically generated files needed for the nodes in the pool
+// A EC2 UserData ready base64 string will be returned.
+func (r *NodePoolTemplateRenderer) prepareCloudInit(nodePool *api.NodePool, values map[string]interface{}) (string, error) {
+	var (
+		keyName, poolKind string
+	)
+	if nodePool.IsMaster() {
+		keyName = masterFilesKMSKey
+		poolKind = "master"
+	} else {
+		keyName = workerFilesKMSKey
+		poolKind = "worker"
+	}
+
+	kmsKey, err := getPKIKMSKey(r.awsAdapter, r.cluster.LocalID, keyName)
 	if err != nil {
 		return "", err
 	}
 
-	values[userDataValuesKey] = renderedUserData
+	renderer := &FilesRenderer{
+		awsAdapter: r.awsAdapter,
+		cluster:    r.cluster,
+		config:     r.config,
+		directory:  poolKind,
+		nodePool:   nodePool,
+	}
+
+	s3Path, err := renderer.RenderAndUploadFiles(values, r.bucketName, kmsKey)
+	if err != nil {
+		return "", err
+	}
+
+	r.logger.Debugf("Uploaded generated files to %s", s3Path)
+	values[s3GeneratedFilesPathValuesKey] = s3Path
+
+	cloudInitContents, err := r.config.NodePoolManifest(nodePool.Profile, cloudInitFileName)
+	if err != nil {
+		return "", err
+	}
+	rendered, err := renderSingleTemplate(cloudInitContents, r.cluster, nodePool, values, r.awsAdapter)
+	if err != nil {
+		return "", err
+	}
 
-	stackManifest, err := p.config.NodePoolManifest(nodePool.Profile, stackFileName)
+	if !r.encodeUserData {
+		return rendered, nil
```

Comment:
What's the use case of this? For testing?

I'm asking because according to the [AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-add-user-data.html) userdata always needs to be base64 encoded, so what's the point of not encoding it?

Reply:
I tried to make karpenter work with this way of encoded userdata and I failed 😅 
So as a workaround, I bypassed the encoding, and that worked fine

we will need to revisit this in the future and enforce the encoding (after learning the right way to do it)
but for now, I did not want to be blocked by it

Reply:
At some point we added the `gzip` part because either the Userdata or the CloudFormation Template size exceeded some limit. I don't know if that's still an issue but just a heads up in case you stumble upon it.

Reply:
Yea, I did stumble upon it, but I could guess it is related to the payload size

In karpenter, userdata is not part of the cloudformation template; rather, it is used by the karpenter controller to create the launch template and then the ec2 node dynamically. Somewhere in this pipeline, it was not able to parse the gzipped-bas64-encoded userdata and was creating corrupted launch templates, so the ec2 node never run

Reply:
Thanks for the explanation.